### DISCUSSION TITLE
Exercise 6: Chandy-Lamport Snapshot Protocol builds a consistent global state.

### DISCUSSION THREAD

**Discussion by dag7dev** [sapienzastudentsnetwork/distributed-systems#31] [Exercises]
**Exercise 6:** Show that the Chandy-Lamport Snapshot Protocol builds a consistent global state.

**Comment by susannacifani** [comment]
I'm not sure about this one.

We know that the Chandy-Lamport protocol makes use of FIFO channels, so let’s assume the channels are FIFO.
If we take a cut $S'$, that is made from a snapshot and an event, and that it's not consistent, we have $e \rightarrow e' \land e' \in S' \Rightarrow e \notin S'$.
We have a snapshot $s1$ taken before $e$ because $e \notin S'$ and we have a snapshot $s2$ after $e'$ because $e' \in S'$.
$s1$ will trigger a take snapshot message that will arrive after $s2$ (and $e'$) and this is impossible because our channels are FIFO.
So we can say that a global state is consistent if $e \rightarrow e' \land e' \in S \Rightarrow e \in S$ and that the Chandy-Lamport algorithm builds a consistent global state.

**Comment by rotiroti** [comment]
Let $e$ and $e'$ two events. We aim to prove that if $e'$ is part of the snapshot $S$ and $e \to e'$, then $e$ should also be in $S$.
Let's demonstrate it by contradiction.

- Assume the contrary: $e' \in S \land e \notin S$.
- This contradiction is immediately evident if $e$ and $e'$ are on the same process because a process's state is captured atomically.
- If $e$ is a send event and $e'$ is a corresponding receive event on another process, then the marker message causing $e'$ to be in the snapshot $S$ must have been after $e$ since the Chandy-Lamport Snapshot Protocol assumes that channels are FIFO. Thus, it means the process of event $e$ must have recorded its state (including $e$) before sending the marker. Therefore, $e$ should be in S, which contradicts our assumption.

We conclude from the contradiction that if $e'$ is in the snapshot $S$ and $e \to e'$, then $e$ must also be in $S$.

**Comment by Tommaso-Sgroi** [comment]
Assume communication's channels are FIFO and network reliable.
We list 3 properties of the snapshot protocol that can show the build of a consistent global state:

1. A process $P_i$ before sending the "take-snapshot" message must deliver its outgoing message, if there is one (FIFO assumption).
2. After receiving the first "take-snapshot" from process $P_j$ message, a process $P_i$ starts recording incoming messages from all channels excluding $P_j$'s one.
3. After receiving another "take-snapshot" message (we have received a "take-snapshot" message yet) from process $P_k$, we stop recording incoming messages from channel $P_k$ and we can add those to the local history.
4. If we receive $n-1$ "take-snapshot" messages we can send our local history.
 
Being $S$ the global state as the merge of processes local histories got from (4), for each event we have $e \rightarrow e' \in S$. $e \in S$ is ensured by (1) and $e' \rightarrow S$ is ensured by (2) and (3).

**Comment by cornflexxx** [comment]
To show the consistency of the global state $\Sigma_s$, we've to show that the corresponding cut is consistent.  
By the design of the protocol the global state is composed of all the last prerecording events, the corresponding  cut $C_s$ is the set of all the prerecording events. To prove the consistency of $C_s$ we have to show that for every $e \in C_s$ (i.e. $e$ is a prerecording event) and $e' \to e$ implies that $e' \in C_s$(that is $e'$ is a prerecording too). Suppose for the sake of contradiction that exists some $e \in C_s$ s.t $e'\to e$ and $e' \not \in C_s$. Rephrased, a postrecording event  ($e'$) causally precedes a prerecording event ($e$), now we can distinguish in two cases:
-  $e'$ and $e$ are two event of the same process, then by assumption $e' \to e$, $e' \not\to e^{\ast}$ (where $e^{\ast}$ is the recording event) and $e \to e^{\ast}$. Using transitivity $e' \to e^{\ast}$ contradiction. 
-  $e'$  is a send event  in some $p_i$ and $e$ is the corresponding receive event in $p_j$, but since $e' \to e$, $p_i$ sent a  "take snapshot" to $p_j$   before $e'$ and assuming FIFO delivery we know "take snapshot" message reaches $p_j$ before $e$, hence $e$  is a postrecording event $\Rightarrow\Leftarrow$.

**Comment by Exyss** [comment]
Even thought the previous solutions all share the same idea, none of them accounts for the fact that $e \to e'$ might be derived through transitivity of the $\to$ relation. The argument below fixes this small issue, preserving the already stated ideas (sadly, this fix requires to be super-formal).

Let $\Sigma$ be the global state derived through the CL protocol. For each process $p_i$, let $e_i^{\*}$ be the recording event (i.e. when the process receives its first "take snapshot" message). Let $C = <h_1^{\*}, \ldots, h_n^{\*}>$ be the cut associated with $\Sigma$. In order to show that $C$ is consistent, we prove the following stronger claim.

Claim: For every chain $e^{(1)} \to \ldots \to e^{(\ell)}$ such that $e^{(\ell)} \in C$ all events of the chain are in $C$.
Proof.
By way of contradiction, suppose that there is at least a chain for which the claim doesn't hold. Let $\mathcal{C} = e^{(1)} \to \ldots \to e^{(\ell)}$ be a chain of minimal length for which the claim doesn't hold. Let $p_i$ be the process of $e^{(\ell)}$. We may assume that at least one relation of the chain comes from send/receive pair, otherwise each event $e^{(j)}$ of the chain is an event of $p_i$ that occurred before $c^{(\ell)}$, concluding that $e^{(j)} \to e_i^{\*}$, hence $e^{(j)} \in h_i^{\*}$ and thus $e^{(j)} \in C$.
Let $t$ be the minimal index of the chain such that $e^{(t)} \to e^{(t+1)}$ comes from the fact that $e^{(t)}$ is a send event in process $p_k$ and $e^{(t+1)}$ is its corresponding receive event in process $p_h$.

Since $e^{(t+1)} \to \ldots \to e^{(\ell)}$ is a chain smaller than $\mathcal{C}$, it must satisfy the property of the claim, concluding that  $e^{(t+1)}, \ldots,  e^{(\ell)} \in C$. Since $e^{(t+1)} \in h_h^{\*}$, it must hold that $e^{(t+1)} \to e_h^{\*}$. We observe that $e^{(t)} \notin C$ must hold, otherwise $e^{(1)}, \ldots,  e^{(t)} \in C$ by the same minimality argument, but this is impossible since at least one event of $\mathcal{C}$ must be outside of $C$.
Since $e^{(t)} \notin C$, it must hold that $e_k^{\*} \to e^{(t)}$. However, this implies that process $p_k$ sent the "take snapshot" message  to every process ($p_h$ included) before sending the message of event $e^{(t)}$. Since the CL protocol uses the FIFO assumption, this  "take snapshot" message must have been received by $p_h$ before of $e^{(t+1)}$, meaning that $e_h^{\*} \to e^{(t+1)}$ and thus raising a contradiction. Therefore, the only possibility is that the claim must hold for every chain.

**Reply by cornflexxx** [reply]
It's not necessary to consider this case: in the proof we're considering only $subsequent$ pair of events, so suppose that $e\to e'$ using transitivity( say through another  event $e^+$, $e^+ \to e'$ and $e \to e^+$) then they can't be subsequent in the run $r$, since $r$ is consistent can have only the pair  $\langle e^+,e'\rangle$.

**Reply by Exyss** [reply]
I don't see why we should consider only subsequent pairs, could you clarify on that?

**Reply by cornflexxx** [reply]
I'm sorry, maybe I was too vague. The goal of the proof is to show that we can build consistent global state (so consistent cut) through CS protocol. Consider a consistent run
  
 $$ r = e_1^{\ast}e_2^{\ast}\dots e_k^{\ast}e_{k+1}e_{k+2}e_{k+3}^{\ast}\dots$$

where all the $e_i^{\ast}$ are in the cut corresponding to the global state built from CS protocol. So the only thing that we need is to pairwise switch events not in the cut with event in the cut in order to get a consistent run $R'$ which reaches the cut (at every step a consistent run has a  corresponding consistent global state). In fact we show that we can safely swap such events maintaining the run consistent ($e^\{*}\not\to e$).

**Reply by Exyss** [reply]
It's a bit more clear now, but I still can't understand your point. From what I grasped, you're considering any consistent run $R$ to then pairwise swap the position of each pair of events $e,e'$ such that $e to e'$, $e' \in R$ and $e \to e'$ in order force that $e \not\to e'$, building a run $R'$ that reaches the global state returned by the CL protocol. If my interpretation is correct, this idea doesn't make sense even without considering transitivity: if $e$ is a send and $e'$ is the corresponding receive, swapping them will give me an inconsistent run.

Moreover... both of the last ideas weren't even mentioned in your solution...?
